### PR TITLE
Ratelimit: Add dynamic metadata to ratelimit actions

### LINKED ISSUE #11593: Proposal: Add dynamic metadata to rate limit actions
**Description**: Currently to pass information from the dynamic metadata to the rate limit actions one would need to use the Lua filter to append a header with the relevant information. Instead, it would be more useful to be able to use the dynamic metadata directly as one of the actions of the rate limiter.

The action can follow the current MetaData schema[1] to get the value: 

```yaml
actions:
  - { dynamic_metadata: { key: “…”, path: [] } }
```
key: The key name of metadata to retrieve the Struct from the metadata. 
path: The path to retrieve the value from the Struct

**Links**:
1. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/metadata/v3/metadata.proto.html

### PR DESCRIPTION
**Pull request by clarakosi**
Commit Message: Modifies ratelimit filter to be able to use information
from the dynamic metadata as one of its actions.

Signed-off-by: Clara Andrew-Wani <candrewwani@gmail.com>

Risk Level:
Testing:  unit tests
Docs Changes: Updated rate limit actions to now include dynamic metadata
Release Notes:  
Fixes #11593

### CODE REVIEW

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -1474,6 +1488,8 @@ message RateLimit {
 
       // Rate limit on the existence of request headers.
       HeaderValueMatch header_value_match = 6;
+
+      DynamicMetaData dynamic_metadata = 7;
```
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:1492]: Can you add a one liner comment above as per other `oneof` alternatives?

Comment (clarakosi) [reply api/envoy/config/route/v3/route_components.proto:1493]: Done

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -103,6 +103,7 @@ New Features
   tracing is not forced.
 * router: add support for RESPONSE_FLAGS and RESPONSE_CODE_DETAILS :ref:`header formatters
   <config_http_conn_man_headers_custom_request_headers>`.
+* router: add support for use of dynamic metadata :ref:`dynamic_metadata <envoy_v3_api_field_config.route.v3.RateLimit.Action.dynamic_metadata>` as a ratelimit action.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:106]: nit: I would probably put the category here as "ratelimit"

File: source/common/router/router_ratelimit.cc [hunk 3]
```diff
@@ -63,12 +70,30 @@ bool RemoteAddressAction::populateDescriptor(
...
+    const envoy::config::route::v3::RateLimit::Action::DynamicMetaData& action)
+    : metadata_key_(action.metadata_key()), descriptor_key_(action.descriptor_key()) {}
+
+bool DynamicMetaDataAction::populateDescriptor(
+    const Router::RouteEntry&, RateLimit::Descriptor& descriptor, const std::string&,
+    const Http::HeaderMap&, const Network::Address::Instance&,
+    const envoy::config::core::v3::Metadata* dynamic_metadata) const {
+  const ProtobufWkt::Value& metadata_value =
+      Envoy::Config::Metadata::metadataValue(dynamic_metadata, metadata_key_);
+  if (metadata_value.kind_case() != ProtobufWkt::Value::kStringValue) {
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.cc:89]: In the docs for the descriptor type can you document this requirement?

File: docs/root/version_history/current.rst (hunk 2 above)
Comment (clarakosi) [reply docs/root/version_history/current.rst]: Done.

File: source/common/router/router_ratelimit.cc (hunk 3 above)
Comment (clarakosi) [reply source/common/router/router_ratelimit.cc:89]: Done.

**Comment by dio** [comment]
@Pchelolo @clarakosi merging master probably would help.

**Comment by mattklein123** [comment]
TSAN failure is a known issue, will merge.

**Pull request merged**